Use alluxio Java API to mount instead of cmd [databricks]

res-life opened the pull request:
Contributes to #6597, #6517

Currently, we use an Alluxio command line to get the mount table and mount an s3 bucket.
This PR introduced the Alluxio client jar to do:
- Get the mount table
- mount an s3 bucket

Pros:
- Avoid parsing the output of `./alluxio fs mount`
- Can remotely access Alluxio in the future. Currently, we assume the Spark driver and Alluxio master are on the same node.
   If the Spark driver and Alluxio master are not on the same node in the future, we can utilize this Java Client to access the remote Alluxio master.
- Can utilize this Java Client to access necessary info, such as if file is cached in issue: #6517
- Can get more readable errors for `get mount table` and `mount bucket` operators.


Cons:
- May introduce incompatible issues in the future.

Signed-off-by: Chong Gao <res_life@163.com>

res-life commented:
Tested in the databricks docker, but not tested the Kratos Env which requires no access and secret keys.

tgravescs commented:
so can you add to description why we are doing this?  Is this just cleanup to get rid of call out of cli?
Is the client pretty compatible with multiple versions or do we know if they break compatibility a lot?

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:
```diff
@@ -204,59 +201,69 @@ object AlluxioUtils extends Logging {
+    // get s3 mount points by alluxio client
+    withResource(alluxio.client.file.FileSystem.Factory.create(conf)) { fs =>
+      val mountTable = fs.getMountTable
+      mountTable.asScala.filter(e => e._2.getUfsType == "s3")
```
ufs type will always be s3 and not s3a or something like that?

this woudl be more readable like:
`      mountTable.asScala.filter{ case (_, mountPoint) => mountPoint.getUfsType == "s3"}
`

res-life replied:
Checked it should be "s3", anyway also added "s3a"

[S3AUnderFileSystem](https://github.com/Alluxio/alluxio/blob/v2.8.0/underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java#L341-L344)
  
public String getUnderFSType() {
    return "s3";
  }

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:
```diff
@@ -538,7 +549,8 @@ object AlluxioUtils extends Logging {
     val hadoopConf = relation.sparkSession.sparkContext.hadoopConfiguration
     val runtimeConf = relation.sparkSession.conf
-    initAlluxioInfo(conf)
+    initAlluxioInfo(conf, relation.sparkSession.sparkContext.hadoopConfiguration,
```
we already have this set to hadoopConf variable above use that. Same thing for runtimeConf

res-life replied:
done

res-life commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:
```diff
@@ -147,43 +153,34 @@ object AlluxioUtils extends Logging {
+          try {
+            val (access_key, secret_key) = getKeyAndSecret(hadoopConf, runtimeConf)
+            val mountPoints = getExistS3MountPoints(conf.getAlluxioUser, access_key, secret_key)
+            mountPoints.foreach { e =>
```
done.

res-life commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:
```diff
@@ -147,43 +153,34 @@ object AlluxioUtils extends Logging {
-          } else {
-            logWarning(s"Failed to run alluxio fs mount $ret")
+          } catch {
+            case e: Exception => logWarning(s"Failed to get alluxio mount table", e)
```
done

res-life commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:
```diff
@@ -204,59 +201,69 @@ object AlluxioUtils extends Logging {
+            mountedBuckets(local_bucket) = remote_path
+          }
+        } catch {
+          case e: Exception =>
```
done

res-life commented:
Tested compatible:
Compile spark rapids against Alluxio version 2.8.0 and the server is 2.8.0 version.
Compile spark rapids against Alluxio version 2.8.0 and the server is 2.8.1 version.
Compile spark rapids against Alluxio version 2.8.1 and the server is 2.8.1 version.

res-life commented:
Build is blocked by: https://github.com/NVIDIA/spark-rapids/issues/6869

res-life commented:
build

res-life commented:
build

res-life commented:
build

res-life commented:
build

res-life commented:
Rebased the code, and re-tested on the databricks cluster.
Fixed the `NoClassDefFoundError` when running UT, see: https://github.com/NVIDIA/spark-rapids/pull/6824/commits/b9d7278c98f65bce1a59c228ddbb435b09386620

res-life commented:
@tgravescs  Help review, thanks.

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioCfgUtils.scala:
```diff
@@ -0,0 +1,71 @@
+ * Utils for the alluxio configurations.
+ * If alluxio feature is disabled, we can tolerate the client jar is absent.
+ * Use this pattern to solve the `NoClassDefFoundError` in `MortgageSparkSuite`:
+ * if (AlluxioCfgUtils.isAlluxioEnabled) { AlluxioCfgUtils does not import any alluxio class
```
this comment seems odd because there is not AlluxioCfgUtils.isAlluxioEnabled function in here.  I'm assuming you mean choose the appropriate enable* function below and call that first before ever calling anything from alluxio client, correct?

res-life replied:
You are right. I updated the comments.

tgravescs commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AlluxioUtils.scala:
```diff
@@ -225,59 +191,71 @@ object AlluxioUtils extends Logging {
       } else {
         throw new RuntimeException(s"Found a same bucket name in $remote_path " +
-          s"and ${mountedBuckets(local_bucket)}")
+            s"and ${mountedBuckets(local_bucket)}")
```
I would expect spacing to be 2 spaces here like it was before

res-life replied:
Done.

res-life commented:
build
